src/Provider/GoogleMapsPlaces/GoogleMapsPlaces.php

```diff
@@ -0,0 +1,410 @@
+            $query['language'] = $reverseQuery->getLocale();
+        }
+
+        $query = $this->applyDataFromQuery($reverseQuery, $query, [
+            'keyword',
+            'type',
+            'name',
+            'minprice',
+            'maxprice',
+            'name',
+            'opennow',
+        ]);
+
+        $requiredParameters = array_filter(array_keys($query), function (string $key) {
+            return in_array($key, ['keyword', 'type', 'name'], true);
+        });
+
+        if (1 !== count($requiredParameters)) {
+            throw new InvalidArgument('One of `type`, `keyword`, `name` is required to be set in the Query data for Reverse Geocoding');
+        }
+
+        return $query;
+    }
+
+    /**
+     * @param Query $query
+     * @param array $request
+     * @param array $keys
+     *
+     * @return array
+     */
+    private function applyDataFromQuery(Query $query, array $request, array $keys)
+    {
+        foreach ($keys as $key) {
+            if (null === $query->getData($key)) {
+                continue;
+            }
+
+            $request[$key] = $query->getData($key);
+        }
+
+        return $request;
+    }
+
+    /**
+     * @param string $url
+     * @param array  $query
+     *
+     * @return AddressCollection
+     */
+    private function fetchUrl(string $url, array $query): AddressCollection
+    {
+        $query['key'] = $this->apiKey;
+
+        $url = sprintf('%s?%s', $url, http_build_query($query));
+
+        $content = $this->getUrlContents($url);
+        $json = $this->validateResponse($url, $content);
+
+        if (empty($json->candidates) && empty($json->results) || 'OK' !== $json->status) {
+            return new AddressCollection([]);
+        }
+
+        $results = [];
+
+        $apiResults = isset($json->results) ? $json->results : $json->candidates;
+
+        foreach ($apiResults as $result) {
+            $builder = new AddressBuilder($this->getName());
+            $this->parseCoordinates($builder, $result);
+
+            if (isset($result->place_id)) {
+                $builder->setValue('id', $result->place_id);
+            }
+
+            /** @var GooglePlace $address */
+            $address = $builder->build(GooglePlace::class);
+            $address = $address->withId($builder->getValue('id'));
+
+            if (isset($result->name)) {
+                $address = $address->withName($result->name);
+            }
+
+            if (isset($result->formatted_address)) {
+                $address = $address->withFormattedAddress($result->formatted_address);
+            }
+
+            if (isset($result->types)) {
+                $address = $address->withResultType($result->types);
+            }
+
+            if (isset($result->icon)) {
+                $address = $address->withIcon($result->icon);
+            }
+
+            if (isset($result->plus_code)) {
+                $address = $address->withPlusCode(new PlusCode(
```

Comment:
Question was around using entities. I created a few for parts of the response, but was not sure if you prefer the raw array (see photos as well below)

Reply:
That's fine I think 👍